sstables/trie: implement BTI node format serialization and traversal

michoecho opened the pull request:
This is the next part in the BTI index project.

Overarching issue: https://github.com/scylladb/scylladb/issues/19191
Previous part: https://github.com/scylladb/scylladb/pull/25154
Next part: implementing a trie cursor (the "set to key, step forwards, step backwards" thing) on top of the `node_reader` added here.

The new code added here is not used for anything yet, but it's posted as a separate PR
to keep things reviewably small.

This part implements the BTI trie node encoding, as described in https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/io/sstable/format/bti/BtiFormat.md#trie-nodes.
It contains the logic for encoding the abstract in-memory `writer_node`s (added in the previous PR)
into the on-disk format, and the logic for traversing the on-disk nodes during a read. 

New functionality, no backporting needed.

avikivity commented on sstables/trie/bti_node_sink.cc:
```diff
@@ -0,0 +1,514 @@
+    // For a given offset bitwidth, contains the index of the leanest node type
+    // in singles[], sparses[] and denses[] which can be used to represent a node with
+    // this offset bitwidth.
+    constexpr uint8_t widths_lookup[] = {
```
static?

avikivity commented on sstables/trie/bti_node_sink.cc:
```diff
@@ -0,0 +1,514 @@
+    // reverse the order (earlier bytes in the transition chain become later nodes in the file)
+    // and add a 0x12 byte before every transition byte.
+    constexpr int s = 16; // Vector size.
+    typedef unsigned char vector2x __attribute__((__vector_size__(s*2))); // Output vector.
```
How does that work? Does the compiler allocate two registers for it?

avikivity commented on test/boost/CMakeLists.txt:
```diff
@@ -264,6 +264,8 @@ add_scylla_test(top_k_test
 add_scylla_test(transport_test
   KIND SEASTAR)
+add_scylla_test(bti_node_sink_test
+  KIND BOOST)
```
Actual test missing

michoecho replied:
Right, I moved the test to a separate commit and forgot to move the `configure.py`/`CMakeLists.txt` additions across commits accordingly.

michoecho replied:
Fixed.

michoecho commented on sstables/trie/bti_node_sink.cc:
```diff
@@ -0,0 +1,514 @@
+    // of constexpr-sized memory ops.
+    {
+        vector1x v;
+        // Load into the `i - i` last bytes of `v`.
```
Comment typo: `s/i-i/i-1/`

michoecho commented on sstables/trie/bti_node_sink.cc:
```diff
@@ -0,0 +1,514 @@
+            s, 15, s, 14, s, 13, s, 12, s, 11, s, 10, s, 9, s, 8,
+            s, 7, s, 6, s, 5, s, 4, s, 3, s, 2, s, 1, s, 0
+        );
+        // Write out the first `2*i` first bytes of `v`.
```
Comment typo: `s/2*i/2*(i-1)/`

avikivity commented:
This is too deep for me to review the algorithms in depth, so I'm reviewing the interfaces and test coverage.

I'd like to see a randomized test that generates random datasets with different selectivity, store that dataset in both a reference structure and the trie, and compare lookups.

michoecho commented on sstables/trie/bti_node_type.hh:
```diff
@@ -0,0 +1,88 @@
+// We want this to be always inlined so that bits_per_pointer is substituted with a constant,
+// and this compiles to a simple load, not to a full-fledged memcpy.
+[[gnu::always_inline]]
+inline uint64_t read_offset(const_bytes sp, int idx, int bits_per_pointer) {
```
`always_inline` isn't dependent on compiler flakiness either.

avikivity replied:
always_inline doesn't guarantee inlining, and inlining doesn't guarantee the compiler will propagate the constant.

michoecho commented:
> This is too deep for me to review the algorithms in depth, so I'm reviewing the interfaces and test coverage.
> 
> I'd like to see a randomized test that generates random datasets with different selectivity, store that dataset in both a reference structure and the trie, and compare lookups.

1. Here we deal with node encodings. This PR only contains the logic for traversing a single node. The logic for traversing a branch of nodes will be in the next PR. If you want a test which writes and reads a whole dataset, then that should wait to until after the next PR. 
2. I wouldn't mind adding a randomized test for node encoding/decoding, it's just I don't have a good idea for how to make a good one. A randomized test just doesn't seem like the right tool here. It's not easy to implement a random generator that would generate dataset shapes that explore various code paths and node types rather than just waste time testing the same code path with different bytes, and not hitting half the code paths even once.
What _maybe_ could be useful here is actual (guided) fuzzing, but we don't have a precedent for that in Scylla. 
3. Are the tests I added (e.g. `bti_node_sink_test.cc::test_body`) not good enough in some aspect?

michoecho commented on utils/cached_file.hh:
```diff
@@ -300,6 +310,11 @@ public:
 
+    using ptr_type = cached_page::ptr_type;
+
+    future<std::pair<ptr_type, bool>> get_shared_page(size_t global_pos, tracing::trace_state_ptr trace_state) {
```
Replaced the pair with a named struct in the newest revision.

michoecho commented:
v1.01: 
- Addressed the cosmetic comments.
- Added a randomized variant of `bti_node_sink_test.cc::test_body`.

@avikivity Anything else, or is this mergeable now?

> #### 🔴 CI State: FAILURE
> release 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18812/) 	❌ 	spider7.cloudius-systems.com
> debug 	[Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/18812/) 	❌ 	spider7.cloudius-systems.com
> #### Build Details:
> 
>     * Duration: 7 hr 18 min

```
[2025-08-06T03:04:32.576Z] + podman login docker.io --username **** --password ****
[2025-08-06T03:04:32.576Z] time="2025-08-06T06:04:31+03:00" level=warning msg="RunRoot is pointing to a path (/run/user/1000/containers) which is not writable. Most likely podman will fail."
[2025-08-06T03:04:32.576Z] Error: creating OCI runtime exit files directory: mkdir /run/user/1000: permission denied
```

https://github.com/scylladb/scylla-pkg/issues/5337

This is the third time in the row the CI has failed for reasons unrelated to the PR, and each reroll takes a day. Shall I keep rerolling or can we break the rule and queue it without passing CI?
